src/Dscp/Educator/Validate.hs

```diff
@@ -2,41 +2,104 @@
+      " got " `mappend` show svfActualPubKey
+    build (SubmissionSignatureMismatch {..}) =
+      "Submission signature mismatch. Submission data " `mappend` show svfSubmissionHash
+      `mappend` " with signature " `mappend` show svfSubmissionSig
+      `mappend` " does not correspond to public key " `mappend` show svfSubmissionSigKey
+
 -- | Validate private block.
--- Block is valid if
--- 1. Header matches body
--- 2. All transactions have valid signatures
--- 3. Transaction signature public key correspond to student public key
-validatePrivateBlk :: PrivateBlock -> Either [Text] ()
+-- Private block is valid iff
+-- 1. Header merkle signature matches computed body merkle signature
+-- 2. All submission in block have valid signatures
+-- 3. All submission signature public keys corresponds to student public key
+--    for given submission
+validatePrivateBlk :: PrivateBlock -> Either [BlockValidationFailure] ()
 validatePrivateBlk pb =
-    let txs = pb ^. pbBody . pbbTxs
-        merkleRoot = getMerkleRoot (fromFoldable txs)
-        headerVer = validateTxHeader (_pbHeader pb) merkleRoot
-        blockValid = verifyGeneric (headerVer <> concatMap validateTx txs)
-    in verResToMonadError toList blockValid
-  where validateTxHeader (PrivateBlockHeader {..}) merkleRoot =
-          [(_pbhBodyProof == merkleRoot, "Header signature missmatch")]
-
-        validateTx (PrivateTx {..}) =
+    let subs =_pbbTxs (_pbBody pb)
+        merkleRoot = getMerkleRoot (fromFoldable subs)
+        headerVer = validateHeader (_pbHeader pb) merkleRoot
+        blockValid = filter ((==False) . fst) (headerVer <> concatMap validateSub subs)
+    in case blockValid of
+         [] -> Right ()
+         x  -> Left  (map snd x)
+  where validateHeader (PrivateBlockHeader {..}) merkleRoot =
+          [(_pbhBodyProof == merkleRoot,
+            MerkleSignatureMismatch { bvfExpectedSig     = _pbhBodyProof
+                                    , bvfActualMerkleSig =  merkleRoot
+                                    })]
+
+        validateSub (PrivateTx {..}) =
           let submission = ssSubmission _ptxSignedSubmission
               witnessKey = _swKey (ssWitness _ptxSignedSubmission)
               witnessSign = _swSig (ssWitness _ptxSignedSubmission)
-              txAddrHash = addrHash (sStudentId submission)
+              subAddrHash = addrHash (sStudentId submission)
               witnessSigValid = verify witnessKey (hash submission) witnessSign
-              witnessKeyValid = hash witnessKey == txAddrHash
-          in [(witnessKeyValid, "Tx public key missmatch")
-             ,(witnessSigValid, "Tx signature not valid")]
+              witnessKeyValid = hash witnessKey == subAddrHash
+          in [(witnessKeyValid,
+               SubmissionInvalid { bvfSubmissionFailure =
+                                     SubmissionPublicKeyMismatch { svfExpectedPubKey = subAddrHash
```

Comment:
Honestly, I recommend you the following indent schema:
```haskell
foo = bar
  where
    barbar = Bar
        { qux = 1
        , other = ...
        }
```
It is easier to read, especially in github diff form, and it is more compact and predictable.